### PR TITLE
Refactored Error handling

### LINKED ISSUE #76: Refactor error handling
https://github.com/mercurius-js/cache/blob/3cebe49a0ce9a8ba613470bc85212851df684dd7/index.js#L139-L182 includes quite a few TODOs. In a few cases we are invoking the original resolver twice.

### PR DESCRIPTION
**Pull request by sameer-coder**
This PR tries to simplify the error handling in the code. I have tried to split the different conditionals into separate functions so that the code is more readable. Added some tests to ensure we cover all cases and coverage is 💯 

Closes #76

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -189,3 +169,47 @@ async function invalidation (invalidate, cache, name, self, arg, ctx, info, resu
...
 }
+
+async function getResultForMutationSubscription ({ self, arg, ctx, info, originalFieldResolver, onError }) {
+  let result = false
+  if (info.operation && (info.operation.operation === 'mutation' || info.operation.operation === 'subscription')) {
+    try {
+      result = await originalFieldResolver(self, arg, ctx, info)
+    } catch (error) {
+      onError(error)
+      return null
```
Comment (mcollina) [inline index.js:180]: This is not correct and a bug in the previous implementation. We should not catch this error nor invoke onError.

File: index.js [hunk 2]
```diff
@@ -189,3 +169,47 @@ async function invalidation (invalidate, cache, name, self, arg, ctx, info, resu
...
+  }
+
+  if (isSkipped) {
+    try {
+      report[name].onSkip()
+      return await originalFieldResolver(self, arg, ctx, info)
+    } catch (error) {
+      onError(error)
+      return null
+    }
```
Comment (mcollina) [inline index.js:204]: This is not correct and a bug in the previous implementation. We should not catch this error nor invoke onError.

**Review by mcollina — COMMENTED** [review]
Good work! There are still a few bugs of the original implementation to fix.

File: index.js [hunk 3]
```diff
@@ -189,3 +169,37 @@ async function invalidation (invalidate, cache, name, self, arg, ctx, info, resu
...
+    result = await originalFieldResolver(self, arg, ctx, info)
+  }
+  return result
+}
+
+async function getResultIfSkipDefined ({ self, arg, ctx, info, skip, policy, name, report, originalFieldResolver, onError }) {
+  const result = false
+  let isSkipped = false
+  try {
+    if ((skip && (await skip(self, arg, ctx, info))) ||
```
Comment (simone-sanfratello) [inline index.js:185]:
```suggestion
  isSkipped = ((skip && (await skip(self, arg, ctx, info))) ||
  (policy && policy.skip && (await policy.skip(self, arg, ctx, info)))) 
```

**Review by simone-sanfratello — CHANGES_REQUESTED** [review]

**Review by simone-sanfratello — APPROVED** [review]

**Pull request merged**